metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java

```diff
@@ -1408,6 +1408,7 @@ void handleDirectoriesOffline(
                     (short) 2));
             log.warn("Directories {} in broker {} marked offline, remaining directories: {}",
                     newOfflineDirs, brokerId, newOnlineDirs);
+            // XXX If the result of this is the broker has no online directories then fence.
```

Comment:
Wouldn't `newOfflineDirs` and `newOnlineDirs` will always be empty until the BrokerRegistrationRecord is updated because they are set operations (intersection, difference) on the list of directories during BrokerRegistration which is an empty set?

Reply:
In the bug case, yes they would always be empty but this is the update for the heartbeat and should only be called if a directory goes offline in a broker. 

However in the bug case there can only be one directory on the broker in 3.6 so the broker should never send a heartbeat that would call this path. Instead since the last directory of the broker went offline (it only has one) the broker should fail. Adding code here would be just an extra protection.